src/MutableMap/MutableMap.dfy

```diff
@@ -0,0 +1,47 @@
+/*******************************************************************************
+*  Copyright by the contributors to the Dafny Project
+*  SPDX-License-Identifier: MIT
+*******************************************************************************/
+
+// RUN: %dafny /compile:0 "%s"
+
+module {:extern "MutableMap"} {:options "/functionSyntax:4"} MutableMap {
+  class {:extern "MutableMap"} MutableMap<K(==),V(==)> {
+    function {:extern "content"} content(): map<K, V>
+      reads this
+
+    constructor {:extern "MutableMap"} ()
+      ensures this.content() == map[]
+
+    method {:extern "put"} Put(k: K, v: V)
+      modifies this
+      ensures this.content() == old(this.content())[k := v]    
+
+    function {:extern "keys"} Keys(): (keys: set<K>)
+      reads this
+      ensures keys == this.content().Keys
+
+    function {:extern "values"} Values(): (values: set<V>)
+      reads this
+      ensures values == this.content().Values
+
+    function {:extern "items"} Items(): (items: set<(K,V)>)
+      reads this
+      ensures items == this.content().Items
+      ensures items == set k | k in this.content().Keys :: (k, this.content()[k])
+
+    function {:extern "find"} Find(k: K): (v: V)
+      reads this
+      requires k in this.Keys()
```

Comment:
While requiring `k in this.Keys()` makes some use cases easier, there's a performance pitfall if user code wants to "get the value associated with `k` or something else if `k` isn't in the map". A user might write e.g.
```dafny
let v = if k in mutableMap.Keys() then Some(mutableMap.Find(k)) else None;
```
in which case the `.Keys()` call creates a new `DafnySet` via a `HashSet` constructor that individually copies each element. The better alternative is to write
```dafny
let v = if k in mutableMap.content() then Some(mutableMap.Find(k)) else None;
```
which creates a new `DafnyMap` but does not copy any elements.

I think it'd be best to add a `function FindIfPresent(k: K): (v: Option<V>)` or something along those lines, both to capture this common use case and to help steer users away from implementing it themselves inefficiently.

(For the same reasons, you might consider adding a `predicate HasKey(k: K)` implemented by each runtime. Then `FindIfPresent` can just combine `HasKey` and the existing presence-requiring `Find`.)